### ISSUE TITLE
"this" is not detected correctly inside a function

### ISSUE THREAD

**Issue by dashaus** [Microsoft/TypeScript#12841]
**TypeScript Version:**  2.1.4

**Code**

```ts
function hi() {
    let link = <HTMLAnchorElement>document.querySelector("#link");
    link.onclick = function(e:Event) {
        this.href = "http://www.typescript.com"; //href not exists as property
    }
}

```

**Expected behavior:**

`this` inside link.onclick function should be a HTMLAnchorElement but it is not.

**Actual behavior:**

`this` is not a HTMLAnchorElement so I don't can to access to properties of a link as "href"

**Comment by aluanhaddad** [comment]
The type definition for onclick looks like this
```ts
onclick: (this: HTMLElement, ev: MouseEvent) => any;
```
If you change it to
```ts
onclick: (this: this, ev: MouseEvent) => any;
```
then you get the expected behavior. Perhaps it should actually be the latter.

**Comment by mhegazy** [comment]
This is a duplicate of https://github.com/Microsoft/TypeScript/issues/12490
